Emoji picker

agzuniverse opened the pull request:
This is a PR that aims to create an Emoji picker that can be used to add reactions to any message.

There was already a file called EmojiPicker.js that seemed to do nothing, so I decided make that into my component.

We can use the 'Add reaction' option in the action sheet to navigate to `EmojiPickerScreen` and tap on a reaction. The screen closes automatically after adding a reaction.

This PR also fixes the incorrect handling of realm emojis by the app.

![emoki_actionsheet](https://user-images.githubusercontent.com/22353313/42627108-aef5008e-85e9-11e8-8ff2-d9c78d702817.png)

![emojipicker](https://user-images.githubusercontent.com/22353313/42627104-aa83a00a-85e9-11e8-95c3-25d313686a49.png)

Fixes #303

jainkuniya commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,102 @@
+
+    return (
+      <Screen title="Add Reaction">
+        <Input
```
A quick comment, it will better to re-use the search property of the screen itself.
As now if screen is scrolled then we need to again come back to top for filtering.

agzuniverse replied:
Done!

borisyankov commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,86 @@
+          keyboardShouldPersistTaps="always"
+          initialNumToRender={20}
+          data={emojis}
+          keyExtractor={item => item}
```
That is probably not correct.
What you need is a value from this `item` object that uniquely identifies it.
The emoji `name` prop is likely good for that.

agzuniverse replied:
`item` in this case is the emoji name, it's not an object.

borisyankov commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,86 @@
+    filter: '',
+  };
+
+  handleInputChange = text => {
```
Add a Flow type for the parameter here.

borisyankov commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,86 @@
+    });
+  };
+
+  addReaction = item => {
```
A type here too.

agzuniverse replied:
Will do!

borisyankov commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,86 @@
+  addReaction = item => {
+    const { auth, dispatch } = this.props;
+    const { messageId } = this.props.navigation.state.params;
+    emojiReactionAdd(auth, messageId, 'unicode_emoji', codePointMap[item], item);
```
Once you've added the types, I am not sure if you'll get an error here, but the parameters look incorrect. If `item` is of type `RealmEmojiType` this code will not be correct.

agzuniverse replied:
`item` is just a string that is the emoji name.

agzuniverse commented:
Updated PR and added flow types.

jainkuniya commented on src/message/messageActionSheet.js:
```diff
@@ -189,6 +194,7 @@ const actionSheetButtons /* ActionSheetButtonType[] */ = [
   { title: 'Reply', onPress: reply, onlyIf: isSentMessage },
   { title: 'Copy to clipboard', onPress: copyToClipboard, onlyIf: isNotDeleted },
   { title: 'Share', onPress: shareMessage, onlyIf: isNotDeleted },
+  { title: 'Add reaction', onPress: addReaction, onlyIf: isNotDeleted },
```
Use `resolveMultiple ` here, we we don't want want this option when message is in outbox.

jainkuniya commented on src/nav/AppNavigator.js:
```diff
@@ -66,6 +67,7 @@ export default StackNavigator(
     notifications: { screen: NotificationsScreen },
     'welcome-help': { screen: WelcomeHelpScreen },
     welcome: { screen: WelcomeScreen },
+    'emoji-picker': { screen: EmojiPickerScreen },
```
Just a tip: Try to maintain alphabetical order. Which helps a lot in searching when list becomes too large.

jainkuniya commented on src/emoji/EmojiPickerScreen.js:
```diff
@@ -0,0 +1,86 @@
+
+  addReaction = (item: string) => {
+    const { auth, dispatch } = this.props;
+    const { messageId } = this.props.navigation.state.params;
```
You can de-structure `navigation` too in the previous line.

jainkuniya commented on src/message/messageActionSheet.js:
```diff
@@ -21,6 +21,7 @@ import {
 } from '../api';
 import { showToast } from '../utils/info';
 import { doNarrow, navigateBack, startEditMessage, deleteOutboxMessage } from '../actions';
+import { navigateToEmojiPicker } from '../nav/navActions';
```
`actions.js` contains all actions including `navActions`. So you can import `navigateToEmojiPicker ` in the previous line (
```
 import { doNarrow, navigateBack, startEditMessage, deleteOutboxMessage } from '../actions';
```)
itself.

agzuniverse commented:
Thanks @jainkuniya , changes made!

borisyankov commented:
I think we want 'Add reaction' at a more prominent place. 
It seems to me this is the most important option and should be at the start.

agzuniverse commented:
@borisyankov done! Add Reaction is now the first option in the action sheet

Will be updating this PR adding support for `realmEmoji` soon.

borisyankov commented on src/emoji/__tests__/emojiSelectors-test.js:
```diff
@@ -17,30 +17,35 @@ describe('getActiveRealmEmoji', () => {
       ],
       realm: {
         emoji: {
-          smile: {
```
Ideally, you would merge this commit with the change that made the tests failing.

agzuniverse replied:
These tests were never failing, it's just that the structure of the objects used in these tests where not the same as the actual objects received from the server. That didn't cause the tests to fail, but I thought it's better to have accurate object strictures in these tests.

The new test I added will fail if the structure is wrong.

borisyankov commented on src/emoji/emojiSelectors.js:
```diff
@@ -27,3 +27,11 @@ export const getActiveRealmEmojiTransformed = createSelector(getActiveRealmEmoji
 });
+
+export const getAllRealmEmojiTransformed = createSelector(getAllRealmEmoji, emojis => {
+  const transformedEmojis = {};
```
Use `reduce`. This will combine all three operators in a single, more atomic operation.
The code will be almost the same. Have a look at many other selectors that use it, if you are unsure how to implement.

borisyankov commented on src/emoji/emojiSelectors.js:
```diff
@@ -27,3 +27,11 @@ export const getActiveRealmEmojiTransformed = createSelector(getActiveRealmEmoji
   return transformedEmojis;
 });
+
+export const getAllRealmEmojiTransformed = createSelector(getAllRealmEmoji, emojis => {
```
I don't think the name is telling me enough. 'transformed' is too generic. Can you come up with a better name?

borisyankov commented on src/emoji/__tests__/emojiSelectors-test.js:
```diff
@@ -71,3 +76,41 @@ describe('getAllRealmEmoji', () => {
   });
 });
+
+describe('getActiveRealmEmojiTransformed and getAllRealmEmojiTransformed', () => {
```
Definitely split each selector in a separate test (and separate `describe` group)

borisyankov commented on src/emoji/__tests__/emojiSelectors-test.js:
```diff
@@ -71,3 +76,41 @@ describe('getAllRealmEmoji', () => {
+    };
+
+    expect(getActiveRealmEmojiTransformed(deepFreeze(state))).toEqual(expectedResult);
+    expect(getAllRealmEmojiTransformed(deepFreeze(state))).toEqual(expectedResult);
```
These should be separate tests. Very, very rarely use two `expect`s in the same test (never is ideal :).
Bundling them means that a single test will pass for many reasons. The goal of a test is to help us pinpoint faster what the issue is.
Testing multiple entities in the same test works against that.

borisyankov replied:
It is OK to reuse input data in order to make the multiple tests you necessarily create shorter and contain less repetition.

agzuniverse replied:
Okay, will split this test into two shortly.

borisyankov replied:
I usually try to provide at least three tests per entity (sometimes much more):
* how it behaves when no data is yet available
* how does it behave with expected data
* what does it do with incorrect data

agzuniverse commented:
Modified the old `getAllRealmEmoji` selector to return the object in the correct way instead of having new selectors just to do that. Also removed tests I wrote for the new selectors.

Now I feel like I need a separate test for the object transforming function, but such a test would be entirely identical to the "get realm emojis with absolute url" test. For now I've just added the transforming part to the description of the test. Thoughts?

agzuniverse commented:
Thanks for the feedback!

I've made the following changes:
* Moved "Created component EmojiPickerScreen" to before "Add EmojiPickerScreen to navigation routes"
* Squashed the "Add reaction added to messages_en" commit with the commit where it was first used and rewrote it's commit message.
* Reordered the last few commits in a way that made more sense.
* Rewrote the commit message of "Modified emojiSelectors to return correct object structure" to be more detailed.
* Rewrote the commit message of "Fix emojiSelectors-tests" to be clear on the changes.

gnprice commented:
I don't understand this remark:
```
    Change type of RealmEmoji id from number to string as emojicode in the
    api functions emojiReactionAdd and emojiReactionRemove is the realm
    emoji id, and this is expected to be a string.
```
How is this commit's change connected to `emojiReactionAdd` etc.?

This `RealmEmojiType` is used in several places -- most notably, it describes part of our Redux state. I think the point is that the values we actually have there under `id` are strings.

gnprice commented:
Merged -- thanks again @agzuniverse ! I cut out the realm emoji support, like I said above.

Also removed some unused imports which the linter warned about. Do be sure to check what `yarn test` says about your code, and fix things; we don't want to accumulate warnings, because then everyone has to learn to ignore them.

Meanwhile I filed #2846 for the issue with realm emoji (and Zulip extra emoji). Let's make those fixes a new PR.
